identity-service/src/awsSNS.js

```diff
@@ -87,31 +86,59 @@ async function publish (message, userId, tx, playSound = true) {
 
   if (formattedMessage) {
     logger.debug('AWS SNS formattedMessage', formattedMessage)
-    PUSH_NOTIFICATIONS_BUFFER.push(formattedMessage)
+    const bufferObj = {
+      metadata: { userId, deviceToken: deviceInfo.deviceToken },
+      notification: formattedMessage
+    }
+    PUSH_NOTIFICATIONS_BUFFER.push(bufferObj)
   } else return null
 }
 
+// Actually send the messages from the buffer to SNS
+// If a device token is invalid attempt to remove it
+//
+// DON'T throw errors in this function because it stops execution,
+// we want it to continue
 async function drainPublishedMessages () {
-  try {
-    // TODO (DM) - batch this. DON'T DO Promise.all()
-    for (let i = 0; i < PUSH_NOTIFICATIONS_BUFFER.length; i += PUSH_NOTIFICATION_SLICE_SIZE) {
-      const pushNotifsSlice = PUSH_NOTIFICATIONS_BUFFER.slice(i, i + PUSH_NOTIFICATION_SLICE_SIZE)
-      await Promise.all(pushNotifsSlice.map((notification) => {
-        publishPromisified(notification)
-      }))
+  for (let bufferObj of PUSH_NOTIFICATIONS_BUFFER) {
+    try {
+      const { notification } = bufferObj
+      await publishPromisified(notification)
+    } catch (e) {
+      if (e && e.code && (e.code === 'EndpointDisabled' || e.code === 'InvalidParameter')) {
+        try {
+          const { deviceToken, userId } = bufferObj.metadata
+          // this notification is not deliverable to this device
+          // remove from deviceTokens table and de-register from AWS
+          const tokenObj = await models.NotificationDeviceToken.findOne({
+            where: {
+              deviceToken,
+              userId
+            }
+          })
+
+          if (tokenObj) {
+            // delete the endpoint from AWS SNS
+            await deleteEndpoint({ EndpointArn: tokenObj.awsARN })
```

Comment:
should we delete on a single failure? what happens if the device is temporarily unavailable like on a flight or something?

Reply:
ah i suppose that hinges on whether SNS marks the endpoint as 'disabled' or not after one failure